tcf_website/views/browse.py

```diff
@@ -213,67 +214,66 @@ def course_instructor(request, course_id, instructor_id):
-            "id": subdept[0],
-            "courses": []
-        }
-
-    for course in courses:
-        course_subdept = course.subdepartment.name
-        course_data = {
-            # autopep8 makes the formatting for this wack, sorry
-            'name': str(course), 'id': course.id,
-            'avg_rating': safe_round(
-                Instructor.average_rating_for_course(
-                    instructor, course)),
-            'avg_difficulty': safe_round(
-                Instructor.average_difficulty_for_course(
-                    instructor, course)),
-            'avg_gpa': safe_round(Instructor.average_gpa_for_course(instructor, course)),
-            'last_taught': str(course.semester_last_taught)}
-        grouped_courses[course_subdept]['courses'].append(course_data)
-
-    return grouped_courses
+    instructor: Instructor = get_object_or_404(Instructor, pk=instructor_id)
+    stats: Dict[str, float] = Instructor.objects\
+        .filter(pk=instructor_id)\
+        .prefetch_related('review_set')\
+        .aggregate(
+            avg_gpa=Avg('courseinstructorgrade__average'),
+            avg_difficulty=Avg('review__difficulty'),
+            avg_rating=(
+                Avg('review__instructor_rating') +
+                Avg('review__enjoyability') +
+                Avg('review__recommendability')
+            ) / 3)
+    course_fields: List[str] = ['name', 'id', 'avg_rating', 'avg_difficulty',
+                                'avg_gpa', 'last_taught']
+    queryset: List[Dict[str, Any]] = Course.objects\
+        .filter(section__instructors=instructor, number__gte=1000)\
+        .prefetch_related('review_set')\
+        .annotate(
+            dept=F('subdepartment__name'),
+            name=Concat(
+                F('subdepartment__mnemonic'),
+                Value(' '),
+                F('number'),
+                Value(' | '),
+                F('title'),
+                output_field=CharField(),
+            ),
+            avg_gpa=Avg('courseinstructorgrade__average',
+                        filter=Q(courseinstructorgrade__instructor=instructor)),
+            avg_difficulty=Avg('review__difficulty',
+                               filter=Q(review__instructor=instructor)),
+            avg_rating=(
+                Avg('review__instructor_rating',
+                    filter=Q(review__instructor=instructor)) +
+                Avg('review__enjoyability',
+                    filter=Q(review__instructor=instructor)) +
+                Avg('review__recommendability',
+                    filter=Q(review__instructor=instructor))
+            ) / 3,
+            last_taught=Concat(
+                F('semester_last_taught__season'),
+                Value(' '),
+                F('semester_last_taught__year'),
+                output_field=CharField(),
+            ),
+    ).values('dept', *course_fields)\
+        .order_by('dept', 'name')
+    courses: Dict[str, List[Dict[str, Any]]] = {}
```

Comment:
So my understanding of this is that the `courses` variable is a `Dict` that has department name strings corresponding to a list of course data (each in its own `Dict`) for every course in that department, whereas `queryset` is a simpler, ungrouped list of course data. For clarity, I think it's better to rename `queryset` to `courses` and `courses` to `grouped_courses` or something like that. I'm not too committed to the name I proposed for `courses`, but I definitely think `queryset` alone doesn't convey enough information.

Reply:
Also, I think there should be whitespace between the last `order_by` line in the block related to `queryset` and the start of where you format `courses`, but I see that you've kept them all together so that might just be a stylistic difference.

Reply:
Ah, yes. I was focusing on the functionality and neglected coming up with good variable names. Also, the `.values()` line does need whitespace before it. Fixed in 001b6592654e68ce1811a4652c6a21c4e4866f46